### ISSUE TITLE
[python package] lightgbm Dataset.save_binary() affects model train

### ISSUE THREAD

**Issue by AlbertoEAF** [microsoft/LightGBM#2520]
Operating System: Linux x86_64 & python3.7 & lgbm 2.3.0 (installed through conda).

**Bug:** 

Calling save_binary on the train_data lgb dataset produces different models, train binary and evaluation results than not calling save_binary on the train data before the train.

**How to reproduce:** 

Saving the lgbm dataset prior to training changes the results of the training:

`

    train_data = lgb.Dataset(
            x_train.to_numpy(), # x_train type: pandas dataframe
            label=y_train, # y_train type: pandas series
            feature_name=FEATURE_NAMES,
            categorical_feature=INDEXED_CATEGORICAL_FEATURES,
            free_raw_data=False,
     )

    val_data = lgb.Dataset(
            x_val.to_numpy(),
            label=y_val,
            feature_name=FEATURE_NAMES,
            categorical_feature=cfg.INDEXED_CATEGORICAL_FEATURES,
            free_raw_data=False,
            reference=train_data,
    )

    ### >> problematic lines - model changes if they are called or not!! ###
    train_data.save_binary("<SOMEWHERE>")
    val_data.save_binary("<SOMEWHERE>")
    ### << problematic lines - model changes if they are called or not!! ###

    model = lgb.train(
        cfg.TRAIN_PARAMS.copy(),
        train_data,
        feval=recall_at_1pc_fpr,
        valid_sets=[val_data],
        num_boost_round=cfg.TRAIN_ITERATIONS,
        feature_name=FEATURE_NAMES,
        categorical_feature=cfg.INDEXED_CATEGORICAL_FEATURES,
        early_stopping_rounds=EARLY_STOPPING_ROUNDS,
    )
`

**Different models** are generated according to wether the "problematic" lines where **save_binary is called** for train and val lgb datasets or not. If one saves the train data after the train again, or doesn't run train at all, the hash of the train data will be consistent before and after train, whilst if one doesn't run train_data.save_binary() before training, the hash of that dataset will change, as well as the resulting model - and subsequent evaluation results.

**What was expected:**

Calling save_binary shouldn't modify the dataset nor resulting models and subsequent results.

**Comment by guolinke** [comment]
The Dataset object is lazy-inited, due to it cannot be changed after inited.
`save_binary` will force the init of Dataset, and then it could not be changed.
Then, the `cfg.TRAIN_PARAMS.copy()` in lgb.train will affect the init of Dataset.
Above all, in your case, you should pass the `cfg.TRAIN_PARAMS.copy()` in lgb.Dataset too.

**Comment by AlbertoEAF** [comment]
Hello,

This does look like a bug @guolinke, could you take a look again?
I've done some tests with passing the train params to the lgb datasets too and this affects the model building, namely the created trees, even when using save_binary all the time, and changing parameters afterwards changes results again.

This is not good as I'm having incoherent results after parameter sweeps, just because in those models I'm not saving the binaries before train, thus the optimal parameter choices changes with having saved or not the binaries prior to train.

Also, passing parameters to the dataset also changes the generated models into yet another model, results of which I cannot conciliate with either version that doesn't save binaries or saves without the parameters. This could make sense with your explanation of freezing the dataset with the save_binary, but setting params in the datasets generates yet another model that is different from those other variants. This gets quite confusing so let me break it down with the test results.

# Tests:

## Test 0 - train != save binary + train => could be related to your explanation

If I have a choice of params P (where seed is fixed and num_threads=1 to be reproducible every time), and features X, loading raw data from disk into lgb datasets (where I only set the features X) and train the model yields different results from adding the save_binary operation before train as in the original message above.

I don't believe this should behave this way because the choice of features and params is not changing between lgb dataset creation and train, thus the outcome should be the same.

## Test 1 - save + train == Save + load binary datasets + train  => EXCELLENT!

This test yields good results consistent with your description.

Whenever I save the binary datasets and only set feature_name and categorical_feature in them before saving,
I get the same model results every time (model with say HASH_A), whether I train once, or load the datasets and from those train again.

Changing the parameter set for lgb.train from **X** to some other subset of features **X_subset** results in the same model with HASH_A after loading the dataset binaries that were saved with the original feature set **X**. This makes sense because you said that save_binary freezes those choices.

## Test 2 - add params to lgb dataset != from all previous variants

Since this test is more involved, I'll detail exactly the used code (except for the val_dataset).

### **v0-reproducible_model.txt**
I first created a model **v0-reproducible** with params P, and called it :
```
train_dataset = lgb.Dataset(..., feature_name=X, categorical_feature=Xcat)
train_dataset.save_binary(path)
booster = lgb.train(train_dataset, params=P)
booster.save_model(...)
```

I got the hash for the model of **HASH_A**.

If I repeat this pipeline I still get **HASH_A** every time.

### **v0-reproducible-with-save-params-in-dataset**
With the same code as before, except for the train dataset creation:
`train_dataset = lgb.Dataset(..., feature_name=X, categorical_feature=Xcat, _params=P_)
`
I get the hash for the model of **HASH_P**.

The resulting models should have the same results as I'm passing exactly the same configuration for the model that I'm passing to the dataset before the save, thus results, even after doing the init of the dataset should be consistent.

I then tested if that was the case by another test.


## Test 3 - changing params after loading the dataset to **P'** changes resulting model yet again

Recalling the last test, if I choose the params P and save them in the dataset and train I get a model with hash **HASH_P**, different from the save_binary without setting the train params even if my choices of parameters don't change.

If I now do a new choice of parameters Pbad (where I set really bad parameters such as max_depth = 2, num_leaves = 2, etc.) and save those parameters in the dataset binaries, I get a model with hash **HASH_Pbad**

Now, if I just load the binaries and train again, I always get a model with **HASH_Pbad**.

**However**, if I now load this binary dataset and train with params=P instead of the Pbad, with which the dataset was saved, the new params are loaded (unlike the feature choices when changed post-saving-datasets which yield always the old results with the saved feature choices) and I get again a model with **HASH_P**.

This is inconsistent with the freeze of the dataset choices.


# Conclusion from tests:
- Save_binary has side-effects - changes outcome model. Doing this step or not, even if the feature selections are the passed to it prior to save & remain consistent with the passed features to train, yields different classification models.
- In code where one always saves the binaries prior to train, selection of features is frozen with save_binary
- In code where one always saves the binaries prior to train, selection of params is **not** frozen with save_binary
- Passing params to datasets prior to save **also** has side-effects, and results in yet another model that cannot be reconciled with any other set of tested procedures
- load_binary works well (assuming you don't change choices of params and features)
